### PR TITLE
fix(gatsby): respect node type owner in deleteNode

### PR DESCRIPTION
**Pull request by frinyvonnick**
## Description

It updates deleteNode action so it respects node type owner

## Related Issues

#11930

### CODE REVIEW

File: packages/gatsby/src/redux/actions.js [hunk 1]
```diff
@@ -403,6 +403,27 @@ actions.deleteNode = (options: any, plugin: Plugin, args: any) => {
   // Always get node from the store, as the node we get as an arg
   // might already have been deleted.
   const node = getNode(id)
+  const pluginName = plugin.name
```
Comment (sidharthachatterjee) [inline packages/gatsby/src/redux/actions.js]: Needs to be wrapped with `if (plugin) {` like other instances in the file

**Comment by frinyvonnick** [comment]
Thank you for your review @sidharthachatterjee 👍 I added a little test 😄

File: packages/gatsby/src/redux/actions.js [hunk 2]
```diff
@@ -403,6 +403,29 @@ actions.deleteNode = (options: any, plugin: Plugin, args: any) => {
...
+    if (node && typeOwners[node.internal.type] !== pluginName)
+      throw new Error(stripIndent`
+          The plugin "${pluginName}" deleted a node of a type owned by another plugin.
+
+          The node type "${node.internal.type}" is owned by "${
+        typeOwners[node.internal.type]
+      }".
+
+          If you copy and pasted code from elsewhere, you'll need to pick a new type name
+          for your new node(s).
```
Comment (pieh) [inline packages/gatsby/src/redux/actions.js]: I know this was copied from `createNode`, but (to me) this note about picking different type name doesn't really work for deleting nodes. Maybe let's just remove lines 417-418

Comment (frinyvonnick) [reply packages/gatsby/src/redux/actions.js]: Yes seems logic, sorry about that 😊

**Comment by frinyvonnick** [comment]
Thank you for your review @pieh 👍 !

**Comment by jessepinho** [comment]
This is a breaking change. Shouldn't it have bumped the major version number?

**Comment by sidharthachatterjee** [comment]
@jessepinho You're absolutely right. This is a breaking change and it seems that we missed that at the time. 

Our sincere apologies for this. We completely understand how frustrating it can be for breaking changes to creep in in minor or patch versions. 

Can you elaborate on the issue you're facing with respect to this? We'd love to help!

**Comment by jessepinho** [comment]
@sidharthachatterjee thanks for understanding! <3 

The issue I'm facing is that I used `onCreateNode` in my `gatsby-node.js` to delete nodes that I deemed invalid — e.g., Contentful articles that had no `slug` field filled in.

There are two problems with simply filtering these nodes out when using GraphQL queries:
1. I have to remember to include that filter in every GraphQL query across the site.
2. I literally _can't_ filter via GraphQL in certain queries. For example, when querying for _author_ entries in Contentful, and then adding the `article` property to the query to get a list of articles that reference this author, there's no way to filter the `article` list for only those that have a `slug` filled in. I'd have to do it client-side after receiving the results of the query, which results in both more client-side work and a larger dataset returned by the query — both of which hurt performance.

In short, I'd like to have only nodes that I consider "valid" to be included in Gatsby's internal node database.

(I am currently considering a PR to `gatsby-source-contentful` to allow a `filterNodes` option or something, which would be a callback that could have a similar effect to what I'm trying to do with `onCreateNode`. But that's a bit of extra work when I already had a working solution. What's the cache issue that this PR fixes?)

**Comment by sidharthachatterjee** [comment]
@jessepinho Ah, that makes sense. Your local plugin should be able to delete anything. I've opened an issue for this in https://github.com/gatsbyjs/gatsby/issues/14870! 

Thank you for all the details.

**Comment by sidharthachatterjee** [comment]
After speaking to @pieh we've come to the conclusion that local plugins _shouldn't_ actually be able to delete anything. This can cause all sorts of issues because of broken references and cause other plugins that _own_ these nodes to break. 

@pieh will elaborate

**Comment by pieh** [comment]
Deleting nodes that you don't own, is asking for trouble. In your example, if that article you deem invalid, would be linked to from some other content - deleting just the article would cause errors. I understand that this is frustrating, but adding this check wasn't just to make people's life harder.

There is undocumented workaround to "impersonate" owner plugin which will allow you to still do this, even if this is something that probably will never be recommended:
```
actions.deleteNode({ node }, { name: `gatsby-source-contentful` })
```

**Comment by jessepinho** [comment]
OK, hm... Would the same problem happen if I were to do my `filterNodes` PR idea above? i.e., make it possible to filter nodes via a callback _before_ they're created by the plugin?
